sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -130,34 +85,32 @@ case class InMemoryTableScanExec(
     val numOutputRows = longMetric("numOutputRows")
     // Using these variables here to avoid serialization of entire objects (if referenced
     // directly) within the map Partitions closure.
-    val relOutput: AttributeSeq = relation.output
-
-    filteredCachedBatches().mapPartitionsInternal { cachedBatchIterator =>
-      // Find the ordinals and data types of the requested columns.
-      val (requestedColumnIndices, requestedColumnDataTypes) =
-        attributes.map { a =>
-          relOutput.indexOf(a.exprId) -> a.dataType
-        }.unzip
+    val relOutput = relation.output
+    val serializer = relation.cacheBuilder.serializer
 
-      // update SQL metrics
-      val withMetrics = cachedBatchIterator.map { batch =>
-        if (enableAccumulatorsForTest) {
-          readBatches.add(1)
-        }
-        numOutputRows += batch.numRows
+    // update SQL metrics
+    val withMetrics = if (enableAccumulatorsForTest) {
```

Comment:
Sure I can make the change I did it because I thought it would be more efficient in the common case when we are not testing, but I could be wrong. My knowledge of the JVM JIT + Scala compilation is a little lacking in this area. I know that with Java a `static final boolean` will be inlined during JIT and checks for that value along with dead code paths are entirely removed.

```
lazy val enableAccumulatorsForTest: Boolean = ...
```
I believe is translated into a function in Scala so I don't think that the JVM JIT is able to do the same thing, but we are creating a closure for the map so that might allow it to inline it.

Either way it is `O(number of batches)` so it should be much less of an issue compared to a row iterator.